remove hard dependency on `forem-redcarpet`

phlipper opened the pull request:
The `redcarpet` gem does not build on JRuby, and this dependency effectively prevents JRuby support for Forem. There are other formatter alternatives, and I have just released [forem-kramdown](https://github.com/phlipper/forem-kramdown) as well. 

I would add a note in the README, but there is already a link to the [Formatters wiki page](https://github.com/radar/forem/wiki/Formatters), where I have added a link to `forem-kramdown`. I would be happy to add an additional note if necessary.

phlipper committed:
remove hard dependency on forem-redcarpet

radar commented:
I agree with this patch, but [our tests do not](https://travis-ci.org/radar/forem/jobs/3890191/#L370). Could you please investigate and submit a fix for it? Maybe just add it as a development dependency instead.

phlipper committed:
update development dependencies:

* upgrade rspec to 2.13
* upgrade capybara to 2.0
* specify versions for dependencies that were
  missing them

phlipper committed:
update Gemfile to support JRuby for testing:

* specify platform specific dependencies
* remove Gemfile.lock and ignore it due to
  platform-specific dependencies

phlipper committed:
update rspec support:

* refactor spec helper to take better advantage of
  the support/ directory
* updates to support newer capybara
* move factories out of support/ to match convention

phlipper committed:
spec/requests -> spec/features for capybara 2.0:

This was mostly a straight move, however there
were some minor updates necessary, mostly due to
capybara 2.0 removing support for Ruby 1.8. These
involved issues with the `#text` method on
`Capybara::Node::Element` objects, as well as some
strange behavior with some specific instances of
`page.should_not have_content` matchers. The
spirit of the tests remain the same, but the
matcher syntax was updated slightly.

phlipper committed:
controller behavior extracted from feature specs

phlipper committed:
swap formatter engines based on ruby version:

* This file had some weird whitespace and this
  diff is best viewed with the `-w` flag

phlipper committed:
add JRuby support to Travis CI test matrix

phlipper commented on spec/features/admin/authentication_spec.rb:
```diff
@@ -16,7 +16,8 @@
       visit admin_root_path
-      page.should_not have_content("Access denied.")
+      # page.should_not have_content("Access denied.")
+      page.html.should_not match("Access denied.")
```
These changes are for Ruby 1.8 / Capybara 2.0 issues with `.should_not` matchers. I haven't tracked down the cause yet, but capybara just [recently made the move to require 1.9+](https://github.com/jnicklas/capybara/commit/82e01429a0b189b9971eb4b843c1037d278a4aff), and these specs were passing on MRI 1.9.3 and JRuby, so I expect we can change these back once Forem requires 1.9+ as well.

phlipper commented on spec/features/admin/forums_spec.rb:
```diff
@@ -44,7 +45,7 @@
 
         flash_alert!("This forum could not be created.")
-        find_field("forum_title").value.should eql("")
+        find_field("forum_title").value.should be_blank
```
These are due to a difference between MRI returning `""` and JRuby returning `nil`. I'm not sure why, but the spirit of the test remains the same.

phlipper commented on spec/features/moderation/approved_users_spec.rb:
```diff
@@ -22,7 +22,11 @@
       visit forum_topic_path(forum, topic)
-      click_link "Reply"
+
+      within("menu") do
```
These updates are due to a behavioral change in Capybara 2.0:

> Ambiguous Matches
> 
> The find method, as well as most actions like click_on, fill_in, etc., now raise an error if more than one element is found. While in Capybara 1.1.2, it would simply select the first matching element, now the matches have to be unambiguous.

phlipper commented on spec/features/posts_spec.rb:
```diff
@@ -10,13 +10,6 @@
       flash_alert!("You must sign in first.")
     end
-
-    it "cannot delete posts" do
```
This example was moved to the controller spec.

phlipper commented on spec/features/topics_spec.rb:
```diff
@@ -13,12 +13,6 @@
       flash_alert!("You must sign in first.")
     end
-
-    it "cannot delete topics" do
```
This example was moved to a controller spec.

phlipper commented:
@radar I got a little sucked in and made some more updates. I've annotated the more "interesting" changes. 

I'm not sure why this is failing some, but not all, Travis builds at the moment. All 3 environments are passing on my machine. I need break for now but I will continue to investigate later.

phlipper commented:
Ok, it looks like the last few failures with the segfaults were due to [an issue with the Travis workers](https://twitter.com/travisci/status/287823719823450112) and have since been resolved. 

The tests now run clean on MRI 1.8/1.9 and JRuby 1.8/1.9.

Thoughts?

parndt commented:
Wow, this is a huge changeset.

parndt commented:
Could you move all un-related changes to a separate branch please? This is too hard to review :(

radar commented:
+1 parndt. Separate out the changes into a different branch so we can pick-and-choose if we please. This PR is for removing forem-redcarpet only.

phlipper commented:
Hi guys, 

I apologize for not getting back to this sooner, I've been wrapped up with a few things. @parndt / @radar, the changes that you indicate are too noisy are there to ensure the current test suite passes as-is. I didn't change any of the existing Forem tests, figuring that would be important. What you are asking requires tests to be removed from Forem, and ideally added to @radar's formatting gem (`forem-kramdown` already has test coverage). 

To confirm, are you asking me to remove the formatting tests from Forem so we can drop the "native platform knowledge" from Forem? Let me know and I will clean this up. Thanks.

anguyen-elc commented:
I have been using this pull in my fork to make it run in JRuby and it's working fine. Please merge this! Thank you.

parndt commented:
There are a lot of commits.. and merge conflicts.. can the commits be squashed and merge conflicts resolved by rebasing against our master please?

phlipper commented:
@parndt I have rebased and squashed some commits. Let me know how this works for you.

parndt commented on .travis.yml:
```diff
@@ -2,13 +2,16 @@ language: ruby
-  - DB=sqlite
+  - CI=true DB=sqlite
+  - CI=true DB=mysql
+  - CI=true DB=postgres
```
`CI=true` is already set by Travis :)

parndt commented on .travis.yml:
```diff
@@ -2,13 +2,16 @@ language: ruby
   - 2.0
   - 1.9.3
   - 1.8.7
+  - jruby
```
This should be `jruby-18mode`

parndt commented:
Why did factories move from `spec/support/factories` to `spec/factories`?

parndt commented:
@phlipper the tests still fail..

phlipper committed:
fix tests:

* ensure email_from_address is set for dummy app
* fix arguments for call to sign_in

parndt commented on .rspec:
```diff
@@ -1 +1 @@
---color
+--color --order rand
```
this is specified in spec_helper

parndt commented on .gitignore:
```diff
@@ -11,3 +11,4 @@ tmp
 spec/dummy
 *.rbc
 *.tmproj
+Gemfile.lock
```
:+1:

parndt commented on forem.gemspec:
```diff
@@ -10,19 +10,19 @@ Gem::Specification.new do |s|
-  s.add_development_dependency 'rspec-rails', '~> 2.6'
-  s.add_development_dependency 'capybara', '1.1.3'
+  s.add_development_dependency 'rspec-rails', '~> 2.13.0'
+  s.add_development_dependency 'capybara', '2.0.2'  # higher requires Ruby 1.9
```
Thanks for documenting this :+1:

parndt commented:
Hi, the tests have been failing on this one for a while and it's not clean to merge due to further changes in the master branch sorry. Hope you're able to fix it up because these changes are great.. I'd love to support jruby and rbx!

parndt commented:
Merged via rebase - thanks.
